kratos/linear_solvers/monotonicity_preserving_solver.h

```diff
@@ -0,0 +1,395 @@
+    * This function tells if the solver requires such data
+    */
+    bool AdditionalPhysicalDataIsNeeded() override
+    {
+        return mpLinearSolver->AdditionalPhysicalDataIsNeeded();
+    }
+
+    /** Some solvers may require a minimum degree of knowledge of the structure of the matrix. To make an example
+     * when solving a mixed u-p problem, it is important to identify the row associated to v and p.
+     * another example is the automatic prescription of rotation null-space for smoothed-aggregation solvers
+     * which require knowledge on the spatial position of the nodes associated to a given dof.
+     * This function is the place to eventually provide such data
+     */
+    void ProvideAdditionalData(
+        SparseMatrixType& rA,
+        VectorType& rX,
+        VectorType& rB,
+        typename ModelPart::DofsArrayType& rdof_set,
+        ModelPart& r_model_part
+    ) override
+    {
+        mpLinearSolver->ProvideAdditionalData(rA,rX,rB,rdof_set,r_model_part);
+    }
+
+    void InitializeSolutionStep (SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
+    {
+        mpLinearSolver->InitializeSolutionStep(rA,rX,rB);
+    }
+
+    /** This function is designed to be called at the end of the solve step.
+     * for example this is the place to remove any data that we do not want to save for later
+    @param rA. System matrix
+    @param rX. Solution vector. it's also the initial guess for iterative linear solvers.
+    @param rB. Right hand side vector.
+    */
+    void FinalizeSolutionStep (SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
+    {
+        mpLinearSolver->FinalizeSolutionStep(rA,rX,rB);
+    }
+
+    /** This function is designed to clean up all internal data in the solver.
+     * Clear is designed to leave the solver object as if newly created.
+     * After a clear a new Initialize is needed
+     */
+    void Clear() override
+    {
+        mpLinearSolver->Clear();
+    }
+
+    /**
+    @param rA. System matrix
+    @param rX. Solution vector. it's also the initial
+    guess for iterative linear solvers.
+    @param rB. Right hand side vector.
+    */
+    bool Solve(SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
+    {
+        if(this->IsNotConsistent(rA, rX, rB))
+            return false;
+
+        double *values_vector = rA.value_data().begin();
+        std::size_t *index1_vector = rA.index1_data().begin();
+        std::size_t *index2_vector = rA.index2_data().begin();
+        const std::size_t matrix_size = rA.size1();
+
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(rA.size1()); i++) {
+            for (std::size_t k = index1_vector[i]; k < index1_vector[i + 1]; k++) {
+                const double value = values_vector[k];
+                if (value > 0.0) {
+                    const int j = index2_vector[k];
+                    if (j > i) {
+                        rA(i,i) += value;
```

Comment:
aside for philipp's comment about using the parallel utils for this, 

the access to the i,j could be optimized a lot: one could store the pointer to A(i,i) for the row you are working on so to avoid accessing to it a lot of times.